subprojects/docs/src/docs/userguide/taskConfiguration.adoc

```diff
@@ -0,0 +1,269 @@
+- Run the https://github.com/gradle/gradle-profiler[gradle-profiler] to measure the difference between a lazy build and a completely eager build. What do you see? Use <<sec:new_task_gradle_profiler_scenario,the scenario file below>>.
+
+[[sec:lazy_task_guidelines]]
+=== Guidelines
+
+[[sec:how_do_i_defer_creation]]
+==== How do I defer task creation?
+
+This feature requires build authors to opt-in by migrating task creation from the api:org.gradle.api.tasks.TaskContainer#create(java.lang.String)[] APIs to the api:org.gradle.api.tasks.TaskContainer#register(java.lang.String)[] APIs. The `register(...)` API registers a task to be created at a later time if and only if the task is needed. The `create(...)` API continues to eagerly create and configure tasks when it is called.
+
+Using the new creation API may not be enough to avoid all task configuration completely. You may need to change other code that configures tasks by name or by type, as explained in the following sections.
+
+[[sec:how_do_i_defer_configuration]]
+==== How do I defer task configuration?
+
+Existing APIs like api:org.gradle.api.DomainObjectCollection#all(org.gradle.api.Action)[] and api:org.gradle.api.DomainObjectCollection#withType(java.lang.Class,org.gradle.api.Action)[] require Gradle to fallback to the old behavior of eagerly creating and configuration tasks. To defer task configuration, you will need to migrate to a new API equivalent. See the <<sec:old_vs_new_configuration_api_overview,table below>> to identify the new API alternative.
+
+[[sec:how_do_i_reference_a_task]]
+==== How do I reference a task without creating/configuring it?
+
+Calling api:org.gradle.api.provider.Provider#get()[] or looking up a task by name with api:org.gradle.api.tasks.TaskCollection#getByName(java.lang.String)[] will cause the task to be created and configured. Methods like api:org.gradle.api.Task#dependsOn(java.lang.Object...)[] and api:org.gradle.api.file.ConfigurableFileCollection#builtBy(java.lang.Object...)[] work with api:org.gradle.api.tasks.TaskProvider[] in the same way as api:org.gradle.api.Task[], so you do not need to unwrap a `Provider` for explicit dependencies to continue to work.
+
+If you are configuring a task by name, you will need to use the new API equivalent. See the <<sec:old_vs_new_configuration_api_overview,table below>> to identify the new API alternative. 
+
+[[sec:how_do_i_get_a_task]]
+==== How to get an instance of a Task?
+
+In the event you still need to get access to a Task instance, you can use api:org.gradle.api.tasks.TaskCollection#named(java.lang.String)[] and call api:org.gradle.api.provider.Provider#get()[].  This will cause the task to be created/configured, but everything should work as it has with the eager APIs. 
+
+Please contact us if you find tricky build logic that requires you to get a Task instance. The intention is to remove the need for the existing eager APIs.
```

Comment:
Missing a link to how to contact us to discuss the tricky build logic. Should it be the forum or GitHub issues?